File: 14.jl
function knothash(ss)
    ls = [[Int(c) for c in ss]; [17, 31, 73, 47, 23]]
    ns = [0:255...]
    skip = 0
    shifts = 0
    for i in 1:64
        for l in ls
            ns = [ns[l:-1:1]; ns[l+1:end]]
            ns = circshift(ns, -l-skip)
            shifts += l+skip
            skip += 1
        end
    end
    ns = circshift(ns,shifts)
    ds = [xor(ns[i:i+15]...) for i in 1:16:256]
    return string([string(x, base=16, pad=2) for x in ds]...)
end

function disk()
    disk = zeros(Int, 128, 128)
    for i in 0:127
        kh = knothash(string("nbysizxe-", i))
        disk[i+1,:] = digits(parse(BigInt, kh, base=16),base=2, pad=128)
    end
    return disk
end

function flood(disk, flooded, row, col)
    pos = disk[row,col]
    flooded[row,col] = 1
    for (r, c) in [(row-1,col), (row+1,col), (row, col-1), (row, col+1)]
        if r in 1:128 && c in 1:128 && disk[r,c] == pos && flooded[r, c] == 0
            flooded = flood(disk, flooded, r, c)
        end
    end
    return flooded
end

function regions(disk)
    regions = 0
    flooded = zeros(Int, size(disk))
    for row in 1:size(disk,1)
        for col in 1:size(disk,2)
            if disk[row,col] == 1 && flooded[row,col] == 0
                regions += 1
                flooded = flood(disk, flooded, row, col)
            end
        end
    end
    return regions
end

println("Part 1: ", sum(disk()))
println("Part 2: ", regions(disk()))

File: 14.py
from functools import reduce

def knothash(cs):
    ls = [ord(c) for c in cs]+[17, 31, 73, 47, 23]
    ns = [x for x in range(256)]
    skip = 0
    shifts = 0
    for i in range(64):
        for l in ls:
            ns = ns[(l-1)::-1] + ns[l:]
            ns = ns[(l+skip)%len(ns):] + ns[:(l+skip)%len(ns)]
            shifts += l+skip
            skip += 1
    ns = ns[-shifts%len(ns):] + ns[:-shifts%len(ns)]
    ds = [reduce(lambda x, y: x^y, list(ns)[i:i+16]) for i in range(0,256,16)]
    return ''.join([ '0'*(2-len(hex(i)[2:])) + hex(i)[2:] for i in ds])

def disk():
    disk = []
    for i in range(128):
        kh = knothash('nbysizxe-'+str(i))
        disk.append([int(c) for c in bin(int(kh,16))[2:].zfill(128)])
    return disk

def flood(disk, flooded, row, col):
    pos = disk[row][col]
    flooded[row][col] = True
    for nrc in [(row-1,col), (row+1,col), (row, col-1), (row, col+1)]:
        if nrc[0] in range(128) and nrc[1] in range(128) and disk[nrc[0]][nrc[1]] == pos and not flooded[nrc[0]][nrc[1]]:
            flooded = flood(disk, flooded, nrc[0], nrc[1])
    return flooded

def regions(disk):
    regions = 0
    flooded = [[False]*len(disk[0]) for x in disk]
    for row in range(len(disk)):
        for col in range(len(disk[0])):
            if disk[row][col] == 1 and not flooded[row][col]:
                regions += 1
                flooded = flood(disk, flooded, row, col)
    return regions

print('Part 1:', sum([sum(x) for x in disk()]))
print('Part 2:', regions(disk()))
